flink-optimizer/src/main/java/org/apache/flink/optimizer/dag/PartitionNode.java

```diff
@@ -90,13 +90,20 @@ public SemanticProperties getSemanticProperties() {
 		private final PartitionMethod pMethod;
 		private final Partitioner<?> customPartitioner;
 		private final DataDistribution distribution;
-		
+		private final Ordering ordering;
+
 		public PartitionDescriptor(PartitionMethod pMethod, FieldSet pKeys, Partitioner<?> customPartitioner, DataDistribution distribution) {
+			this(pMethod, pKeys, null, customPartitioner, distribution);
+		}
+
+		public PartitionDescriptor(PartitionMethod pMethod, FieldSet pKeys, Ordering ordering, Partitioner<?>
+				customPartitioner, DataDistribution distribution) {
 			super(pKeys);
-			
+
 			this.pMethod = pMethod;
 			this.customPartitioner = customPartitioner;
 			this.distribution = distribution;
+			this.ordering = ordering;
```

Comment:
Can you add a check that the `ordering` is valid (number of orders == number of keys. Order keys == keys)?